File: backend-node-archive/src/services/flightPlanDatabaseService.ts
/**
 * flightPlanDatabaseService.ts — Server-side Flight Plan Database API client.
 *
 * API documentation: https://flightplandatabase.com/dev/api
 *
 * Strategy:
 *   1. GET /search/plans?fromICAO=X&toICAO=Y&limit=5&sort=popularity
 *      → metadata list (no full waypoints)
 *   2. GET /plan/{id} for the top result → full route.nodes
 *   3. Remaining results become alternatives (metadata only)
 *
 * Aircraft type and cruising altitude are NOT supported query parameters on the
 * FPD API and are therefore NOT forwarded. They are used locally for enrichment.
 *
 * Results are cached in memory (simple Map) for the configured TTL.
 * The cache is intentionally lost on process restart.
 */

import { FPD_API_KEY, FPD_BASE_URL, FPD_CACHE_TTL_MINUTES } from '../config/env'
import { log } from '../utils/logger'

// ── Error types ────────────────────────────────────────────────────────────────

export type FpdErrorKind = 'configuration_missing' | 'http' | 'network' | 'no_results'

export class FlightPlanDatabaseError extends Error {
  public readonly kind: FpdErrorKind
  public readonly status?: number

  constructor(kind: FpdErrorKind, message: string, status?: number) {
    super(message)
    this.name = 'FlightPlanDatabaseError'
    this.kind = kind
    this.status = status
  }
}

// ── Result types ───────────────────────────────────────────────────────────────

export interface FpdNode {
  type: string    // APT, FIX, VOR, NDB
  ident: string
  name: string | null
  lat: number
  lon: number
}

export interface FpdPlan {
  id: number
  fromIcao: string
  toIcao: string
  distanceNm: number
  maxAltitude: number
  notes: string | null
  updatedAt: string | null
  popularity: number
  nodes: FpdNode[]
}

export interface FpdPlanSummary {
  id: number
  distanceNm: number
  notes: string | null
  updatedAt: string | null
  popularity: number
}

export interface FpdRouteResult {
  selected: FpdPlan
  alternatives: FpdPlanSummary[]
}

// ── Simple in-memory cache ─────────────────────────────────────────────────────

interface CacheEntry<T> {
  value: T
  expiresAt: number
}

const cache = new Map<string, CacheEntry<FpdRouteResult>>()

function getCached(key: string): FpdRouteResult | null {
  const entry = cache.get(key)
  if (!entry) return null
  if (Date.now() > entry.expiresAt) {
    cache.delete(key)
    return null
  }
  return entry.value
}

function setCached(key: string, value: FpdRouteResult): void {
  const ttlMs = Math.max(1, FPD_CACHE_TTL_MINUTES) * 60 * 1000
  cache.set(key, { value, expiresAt: Date.now() + ttlMs })
}

// ── HTTP helper ────────────────────────────────────────────────────────────────

async function callFpdApi<T>(url: string): Promise<T> {
  if (!FPD_API_KEY) {
    throw new FlightPlanDatabaseError(
      'configuration_missing',
      'FLIGHT_PLAN_DATABASE_API_KEY is not configured on this server.',
    )
  }

  const credential = Buffer.from(`${FPD_API_KEY}:`).toString('base64')

  let response: Response
  try {
    response = await fetch(url, {
      method: 'GET',
      headers: {
        Authorization: `Basic ${credential}`,
        Accept: 'application/json',
        'User-Agent': 'FlightPlanner/0.1.0',
      },
    })
  } catch (error) {
    const msg = error instanceof Error ? error.message : String(error)
    log('ERROR', `FPD network error for ${url}: ${msg}`)
    throw new FlightPlanDatabaseError('network', 'Unable to reach Flight Plan Database (network error).')
  }

  if (!response.ok) {
    log('WARN', `FPD HTTP ${response.status} for ${url}`)
    throw new FlightPlanDatabaseError('http', `Flight Plan Database returned HTTP ${response.status}.`, response.status)
  }

  return (await response.json()) as T
}

// ── FPD API raw types ──────────────────────────────────────────────────────────

interface FpdApiPlan {
  id: number
  fromICAO?: string
  toICAO?: string
  distance?: number
  maxAltitude?: number
  notes?: string | null
  updatedAt?: string | null
  popularity?: number
  route?: {
    nodes?: Array<{
      type?: string
      ident?: string
      name?: string | null
      lat?: number
      lon?: number
    }>
  }
}

// ── Public API ─────────────────────────────────────────────────────────────────

/**
 * Searches for IFR flight plans between two airports.
 * Returns the best match with full waypoints plus alternative summaries.
 */
export async function searchRoutes(departure: string, destination: string): Promise<FpdRouteResult> {
  // Check config first (before cache)
  if (!FPD_API_KEY) {
    throw new FlightPlanDatabaseError(
      'configuration_missing',
      'FLIGHT_PLAN_DATABASE_API_KEY is not configured on this server.',
    )
  }

  const cacheKey = `fpd:${departure.toUpperCase()}:${destination.toUpperCase()}:IFR`
  const cached = getCached(cacheKey)
  if (cached) {
    log('INFO', `FPD cache hit for ${departure}-${destination}`)
    return cached
  }

  // Step 1: Search for plans
  const searchUrl =
    `${FPD_BASE_URL}/search/plans` +
    `?fromICAO=${encodeURIComponent(departure)}` +
    `&toICAO=${encodeURIComponent(destination)}` +
    `&limit=5&sort=popularity`

  log('INFO', `FPD search: ${departure} → ${destination}`)
  const summaries = await callFpdApi<FpdApiPlan[]>(searchUrl)

  if (!summaries || summaries.length === 0) {
    throw new FlightPlanDatabaseError('no_results', `No IFR flight plans found between ${departure} and ${destination}.`)
  }

  // Step 2: Fetch full plan for top result
  const topId = summaries[0].id
  log('INFO', `FPD fetching full plan ${topId}`)
  const fullPlan = await callFpdApi<FpdApiPlan>(`${FPD_BASE_URL}/plan/${topId}`)

  if (!fullPlan.route?.nodes || fullPlan.route.nodes.length === 0) {
    throw new FlightPlanDatabaseError('no_results', `Flight plan ${topId} returned no waypoints.`)
  }

  // Step 3: Map to internal model
  const selected: FpdPlan = {
    id: fullPlan.id,
    fromIcao: fullPlan.fromICAO ?? departure,
    toIcao: fullPlan.toICAO ?? destination,
    distanceNm: fullPlan.distance ?? 0,
    maxAltitude: fullPlan.maxAltitude ?? 0,
    notes: fullPlan.notes ?? null,
    updatedAt: fullPlan.updatedAt ?? null,
    popularity: fullPlan.popularity ?? 0,
    nodes: fullPlan.route.nodes.map((n) => ({
      type: n.type ?? '',
      ident: n.ident ?? '',
      name: n.name ?? null,
      lat: n.lat ?? 0,
      lon: n.lon ?? 0,
    })),
  }

  const alternatives: FpdPlanSummary[] = summaries.slice(1).map((s) => ({
    id: s.id,
    distanceNm: s.distance ?? 0,
    notes: s.notes ?? null,
    updatedAt: s.updatedAt ?? null,
    popularity: s.popularity ?? 0,
  }))

  const result: FpdRouteResult = { selected, alternatives }
  setCached(cacheKey, result)

  log('INFO', `FPD found ${summaries.length} plan(s) for ${departure}-${destination}, top id=${topId}`)
  return result
}

File: backend/FlightPlanner.Api/Services/FlightPlanDatabaseService.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using FlightPlanner.Api.Options;
using Microsoft.Extensions.Caching.Memory;

namespace FlightPlanner.Api.Services;

/// <summary>
/// Calls the Flight Plan Database API (https://flightplandatabase.com) to search
/// for IFR routes between two airports. Results are cached in-memory for the
/// configured TTL (default: 30 minutes) to conserve API quota.
///
/// API documentation: https://flightplandatabase.com/dev/api
///
/// Strategy:
///   1. GET /search/plans?fromICAO=X&amp;toICAO=Y&amp;limit=5&amp;sort=popularity
///      → returns up to 5 plan summaries (no full waypoints)
///   2. GET /plan/{id} for the top result → returns full waypoints (route.nodes)
///   3. Remaining search results become alternatives (metadata only)
///
/// Aircraft type and cruising altitude are NOT query parameters supported by the
/// FPD API. They are accepted by our internal endpoint for client-side enrichment
/// and future extensibility, but are not forwarded to FPD.
/// </summary>
public sealed class FlightPlanDatabaseService : IFlightPlanDatabaseService
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    private readonly HttpClient _httpClient;
    private readonly FlightPlanDatabaseOptions _options;
    private readonly IMemoryCache _cache;
    private readonly ILogger<FlightPlanDatabaseService> _logger;

    public FlightPlanDatabaseService(
        HttpClient httpClient,
        FlightPlanDatabaseOptions options,
        IMemoryCache cache,
        ILogger<FlightPlanDatabaseService> logger)
    {
        _httpClient = httpClient;
        _options = options;
        _cache = cache;
        _logger = logger;
    }

    /// <inheritdoc />
    public async Task<FpdRouteResult> SearchRoutesAsync(
        string departure,
        string destination,
        CancellationToken cancellationToken = default)
    {
        // ── Config validation ────────────────────────────────────────────────
        if (string.IsNullOrWhiteSpace(_options.ApiKey))
        {
            throw new FlightPlanDatabaseException(
                FpdErrorKind.ConfigurationMissing,
                "FlightPlanDatabase:ApiKey is not configured on this server. " +
                "Run: dotnet user-secrets set \"FlightPlanDatabase:ApiKey\" \"YOUR_KEY\"");
        }

        // ── Cache lookup ─────────────────────────────────────────────────────
        var cacheKey = $"fpd:{departure.ToUpperInvariant()}:{destination.ToUpperInvariant()}:IFR";
        if (_cache.TryGetValue(cacheKey, out FpdRouteResult? cached) && cached is not null)
        {
            _logger.LogInformation("FPD cache hit for {Dep}-{Arr}", departure, destination);
            return cached;
        }

        // ── Step 1: Search for plans ──────────────────────────────────────────
        var searchUrl = $"{_options.BaseUrl}/search/plans" +
                        $"?fromICAO={Uri.EscapeDataString(departure)}" +
                        $"&toICAO={Uri.EscapeDataString(destination)}" +
                        $"&limit=5&sort=popularity";

        _logger.LogInformation("FPD search: {Dep} → {Arr}", departure, destination);
        var summaries = await CallApiAsync<List<FpdApiPlan>>(searchUrl, cancellationToken);

        if (summaries is null || summaries.Count == 0)
        {
            throw new FlightPlanDatabaseException(
                FpdErrorKind.NoResults,
                $"No IFR flight plans found between {departure} and {destination}.");
        }

        // ── Step 2: Fetch full waypoints for the top plan ─────────────────────
        var topSummary = summaries[0];
        var planUrl = $"{_options.BaseUrl}/plan/{topSummary.Id}";

        _logger.LogInformation("FPD fetching full plan {Id}", topSummary.Id);
        var fullPlan = await CallApiAsync<FpdApiPlan>(planUrl, cancellationToken);

        if (fullPlan?.Route?.Nodes is null || fullPlan.Route.Nodes.Count == 0)
        {
            throw new FlightPlanDatabaseException(
                FpdErrorKind.NoResults,
                $"Flight plan {topSummary.Id} returned no waypoints.");
        }

        // ── Step 3: Map to internal model ─────────────────────────────────────
        var selected = new FpdPlan
        {
            Id          = fullPlan.Id,
            FromIcao    = fullPlan.FromIcao ?? departure,
            ToIcao      = fullPlan.ToIcao   ?? destination,
            DistanceNm  = fullPlan.Distance,
            MaxAltitude = fullPlan.MaxAltitude,
            Notes       = fullPlan.Notes,
            UpdatedAt   = fullPlan.UpdatedAt,
            Popularity  = fullPlan.Popularity,
            Nodes       = fullPlan.Route.Nodes
                .Select(n => new FpdNode
                {
                    Type  = n.Type  ?? string.Empty,
                    Ident = n.Ident ?? string.Empty,
                    Name  = n.Name,
                    Lat   = n.Lat,
                    Lon   = n.Lon,
                })
                .ToList(),
        };

        var alternatives = summaries
            .Skip(1)
            .Select(s => new FpdPlanSummary
            {
                Id         = s.Id,
                DistanceNm = s.Distance,
                Notes      = s.Notes,
                UpdatedAt  = s.UpdatedAt,
                Popularity = s.Popularity,
            })
            .ToList();

        var result = new FpdRouteResult { Selected = selected, Alternatives = alternatives };

        // ── Cache the result ──────────────────────────────────────────────────
        var ttl = TimeSpan.FromMinutes(Math.Max(1, _options.CacheTtlMinutes));
        _cache.Set(cacheKey, result, new MemoryCacheEntryOptions { AbsoluteExpirationRelativeToNow = ttl });

        _logger.LogInformation(
            "FPD found {Count} plan(s) for {Dep}-{Arr}, top plan id={Id}",
            summaries.Count, departure, destination, selected.Id);

        return result;
    }

    /// <inheritdoc />
    public async Task<FpdPlan> FetchPlanByIdAsync(int planId, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(_options.ApiKey))
        {
            throw new FlightPlanDatabaseException(
                FpdErrorKind.ConfigurationMissing,
                "FlightPlanDatabase:ApiKey is not configured on this server.");
        }

        var planUrl  = $"{_options.BaseUrl}/plan/{planId}";
        var fullPlan = await CallApiAsync<FpdApiPlan>(planUrl, cancellationToken);

        if (fullPlan?.Route?.Nodes is null || fullPlan.Route.Nodes.Count == 0)
        {
            throw new FlightPlanDatabaseException(
                FpdErrorKind.NoResults,
                $"Flight plan {planId} returned no waypoints.");
        }

        return new FpdPlan
        {
            Id          = fullPlan.Id,
            FromIcao    = fullPlan.FromIcao ?? string.Empty,
            ToIcao      = fullPlan.ToIcao   ?? string.Empty,
            DistanceNm  = fullPlan.Distance,
            MaxAltitude = fullPlan.MaxAltitude,
            Notes       = fullPlan.Notes,
            UpdatedAt   = fullPlan.UpdatedAt,
            Popularity  = fullPlan.Popularity,
            Nodes       = fullPlan.Route.Nodes
                .Select(n => new FpdNode
                {
                    Type  = n.Type  ?? string.Empty,
                    Ident = n.Ident ?? string.Empty,
                    Name  = n.Name,
                    Lat   = n.Lat,
                    Lon   = n.Lon,
                })
                .ToList(),
        };
    }

    // ── Private helpers ────────────────────────────────────────────────────────

    private async Task<T?> CallApiAsync<T>(string url, CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, url);

        // HTTP Basic Auth: API key as username, empty password
        var credential = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{_options.ApiKey}:"));
        request.Headers.Authorization = new AuthenticationHeaderValue("Basic", credential);
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

        HttpResponseMessage response;
        try
        {
            response = await _httpClient.SendAsync(request, cancellationToken);
        }
        catch (OperationCanceledException)
        {
            throw;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "FPD network error for {Url}", url);
            throw new FlightPlanDatabaseException(
                FpdErrorKind.Network,
                "Unable to reach Flight Plan Database (network error). Please try again later.");
        }

        if (!response.IsSuccessStatusCode)
        {
            _logger.LogWarning("FPD HTTP {Status} for {Url}", (int)response.StatusCode, url);
            throw new FlightPlanDatabaseException(
                FpdErrorKind.Http,
                $"Flight Plan Database returned HTTP {(int)response.StatusCode}.",
                (int)response.StatusCode);
        }

        var body = await response.Content.ReadAsStringAsync(cancellationToken);
        return JsonSerializer.Deserialize<T>(body, JsonOptions);
    }

    // ── FPD API deserialization models (internal only) ─────────────────────────

    private sealed class FpdApiPlan
    {
        [JsonPropertyName("id")]          public int Id { get; set; }
        [JsonPropertyName("fromICAO")]    public string? FromIcao { get; set; }
        [JsonPropertyName("toICAO")]      public string? ToIcao { get; set; }
        [JsonPropertyName("distance")]    public double Distance { get; set; }
        [JsonPropertyName("maxAltitude")] public int MaxAltitude { get; set; }
        [JsonPropertyName("notes")]       public string? Notes { get; set; }
        [JsonPropertyName("updatedAt")]   public string? UpdatedAt { get; set; }
        [JsonPropertyName("popularity")]  public double Popularity { get; set; }
        [JsonPropertyName("route")]       public FpdApiRoute? Route { get; set; }
    }

    private sealed class FpdApiRoute
    {
        [JsonPropertyName("nodes")] public List<FpdApiNode>? Nodes { get; set; }
    }

    private sealed class FpdApiNode
    {
        [JsonPropertyName("type")]  public string? Type { get; set; }
        [JsonPropertyName("ident")] public string? Ident { get; set; }
        [JsonPropertyName("name")]  public string? Name { get; set; }
        [JsonPropertyName("lat")]   public double Lat { get; set; }
        [JsonPropertyName("lon")]   public double Lon { get; set; }
    }
}
